File: java/src/test/java/it/giannibombelli/wsc2026/payment/infrastructure/PaymentRepositoryTest.java
package it.giannibombelli.wsc2026.payment.infrastructure;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.domain.payment.Payment;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentStatus;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentRepository;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import javax.sql.DataSource;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Verifica il contratto del repository senza ripetere le regole di business di dominio.
 */
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class PaymentRepositoryTest {
    private PaymentRepository repository;

    @BeforeAll
    void setUp() {
        final DataSource dataSource = DatabaseSetup.initializeFileDb("payment", getClass().getSimpleName());
        repository = new SqlitePaymentRepository(dataSource);
    }

    @Nested
    class Save {
        @Test
        void shouldPersistNewPayment() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            String clientReference = UUID.randomUUID().toString();
            Money amount = new Money(new BigDecimal("50.00"));
            Instant requestedAt = Instant.parse("2026-06-07T10:00:00Z");
            Payment original = Payment.request(paymentId, new ClientReference(clientReference), amount, new Timestamp(requestedAt));

            repository.save(original);
            Optional<Payment> reloaded = repository.findById(paymentId);

            assertThat(reloaded).isPresent();
            Payment found = reloaded.get();
            assertThat(found.id()).isEqualTo(paymentId);
            assertThat(found.clientReference().value()).isEqualTo(clientReference);
            assertThat(found.amount()).isEqualTo(amount);
            assertThat(found.status()).isEqualTo(PaymentStatus.REQUESTED);
            assertThat(found.requestedAt().value()).isEqualTo(requestedAt);
        }
    }

    @Nested
    class FindById {
        @Test
        void shouldReturnEmptyWhenNotFound() {
            PaymentId nonExistentId = EntityId.generate(PaymentId::new);

            Optional<Payment> reloaded = repository.findById(nonExistentId);

            assertThat(reloaded).isEmpty();
        }
    }

    @Nested
    class FindByClientReference {
        @Test
        void shouldReturnPaymentWhenClientReferenceMatches() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            String clientReference = UUID.randomUUID().toString();
            Money amount = new Money(new BigDecimal("25.00"));
            Payment payment = Payment.request(paymentId, new ClientReference(clientReference), amount, new Timestamp(Instant.now()));
            repository.save(payment);

            Optional<Payment> found = repository.findByClientReference(new ClientReference(clientReference));

            assertThat(found).isPresent();
            assertThat(found.get().id()).isEqualTo(paymentId);
        }

        @Test
        void shouldReturnEmptyWhenClientReferenceDoesNotMatch() {
            Optional<Payment> found = repository.findByClientReference(new ClientReference("non-existing-reference"));

            assertThat(found).isEmpty();
        }
    }

    @Nested
    class FindAllRequestedAndProcessingBefore {
        @Test
        void shouldReturnPaymentsRequestedBeforeThreshold() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            Instant requestedAt = Instant.parse("2026-06-07T10:00:00Z");
            Payment payment = Payment.request(paymentId, new ClientReference(UUID.randomUUID().toString()), new Money(new BigDecimal("10.00")), new Timestamp(requestedAt));
            repository.save(payment);

            List<Payment> found = repository.findAllRequestedAndProcessingBefore(new Timestamp(Instant.parse("2026-06-08T10:00:00Z")));

            assertThat(found).extracting(Payment::id).contains(paymentId);
        }

        @Test
        void shouldNotReturnPaymentsRequestedAfterThreshold() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            Instant requestedAt = Instant.parse("2026-06-07T10:00:00Z");
            Payment payment = Payment.request(paymentId, new ClientReference(UUID.randomUUID().toString()), new Money(new BigDecimal("10.00")), new Timestamp(requestedAt));
            repository.save(payment);

            List<Payment> found = repository.findAllRequestedAndProcessingBefore(new Timestamp(Instant.parse("2026-06-06T10:00:00Z")));

            assertThat(found).extracting(Payment::id).doesNotContain(paymentId);
        }
    }
}

File: typescript/tests/payment/infrastructure/paymentRepository.test.ts
import { beforeAll, describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { ClientReference } from '@/common/domain/primitive/clientReference.js';
import { Money } from '@/common/domain/primitive/money.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { Payment } from '@/payment/domain/payment/payment.js';
import { PaymentId } from '@/payment/domain/payment/paymentId.js';
import { PaymentStatus } from '@/payment/domain/payment/paymentStatus.js';
import { SqlitePaymentRepository } from '@/payment/infrastructure/sqlitePaymentRepository.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import { DatabaseSetup } from '../../testsupport/databaseSetup.js';

describe('PaymentRepository', () => {
  let repository: SqlitePaymentRepository;

  beforeAll(() => {
    const database = DatabaseSetup.initializeFileDb('payment', 'PaymentRepositoryTest');
    repository = new SqlitePaymentRepository(database);
  });

  describe('save', () => {
    it('should persist new payment', () => {
      const paymentId = generateId((value) => new PaymentId(value));
      const clientReference = crypto.randomUUID();
      const amount = new Money(50);
      const requestedAt = new Timestamp(new Date('2026-06-07T10:00:00.000Z'));
      const original = Payment.request(paymentId, new ClientReference(Uuid.fromString(clientReference)), amount, requestedAt);

      repository.save(original);
      const reloaded = repository.findById(paymentId);

      expect(reloaded).not.toBeNull();
      expect(reloaded!.id()).toEqual(paymentId);
      expect(reloaded!.clientReference().toString()).toEqual(clientReference);
      expect(reloaded!.amount()).toEqual(amount);
      expect(reloaded!.status()).toEqual(PaymentStatus.REQUESTED);
      expect(reloaded!.requestedAt().value).toEqual(requestedAt.value);
    });
  });

  describe('findById', () => {
    it('should return empty when not found', () => {
      const nonExistentId = generateId((value) => new PaymentId(value));

      const reloaded = repository.findById(nonExistentId);

      expect(reloaded).toBeNull();
    });
  });

  describe('findByClientReference', () => {
    it('should return payment when client reference matches', () => {
      const paymentId = generateId((value) => new PaymentId(value));
      const clientReference = crypto.randomUUID();
      const amount = new Money(25);
      const payment = Payment.request(
        paymentId,
        new ClientReference(Uuid.fromString(clientReference)),
        amount,
        Timestamp.now()
      );
      repository.save(payment);

      const found = repository.findByClientReference(new ClientReference(Uuid.fromString(clientReference)));

      expect(found).not.toBeNull();
      expect(found!.id()).toEqual(paymentId);
    });

    it('should return empty when client reference does not match', () => {
      const found = repository.findByClientReference(new ClientReference(Uuid.fromString('00000000-0000-0000-0000-000000000001')));

      expect(found).toBeNull();
    });
  });

  describe('findAllRequestedAndProcessingBefore', () => {
    it('should return payments requested before threshold', () => {
      const paymentId = generateId((value) => new PaymentId(value));
      const requestedAt = new Timestamp(new Date('2026-06-07T10:00:00.000Z'));
      const payment = Payment.request(
        paymentId,
        new ClientReference(Uuid.fromString(crypto.randomUUID())),
        new Money(10),
        requestedAt
      );
      repository.save(payment);

      const found = repository.findAllRequestedAndProcessingBefore(
        new Timestamp(new Date('2026-06-08T10:00:00.000Z'))
      );

      expect(found.map((p) => p.id().value.value)).toContain(paymentId.value.value);
    });

    it('should not return payments requested after threshold', () => {
      const paymentId = generateId((value) => new PaymentId(value));
      const requestedAt = new Timestamp(new Date('2026-06-07T10:00:00.000Z'));
      const payment = Payment.request(
        paymentId,
        new ClientReference(Uuid.fromString(crypto.randomUUID())),
        new Money(10),
        requestedAt
      );
      repository.save(payment);

      const found = repository.findAllRequestedAndProcessingBefore(
        new Timestamp(new Date('2026-06-06T10:00:00.000Z'))
      );

      expect(found.map((p) => p.id().value)).not.toContain(paymentId.value);
    });
  });
});
